### PR TITLE
Part of metadata persistence.

### PR DESCRIPTION
**Pull request by ActivePeter**
## Related enhancement issue
https://github.com/datenlord/datenlord/issues/360

TODO:
- left part: flush & test
- https://github.com/datenlord/datenlord/issues/364

### CODE REVIEW

File: scripts/datenlord_perf_test.sh [hunk 1]
```diff
@@ -21,7 +21,7 @@ FIRST_NODE=$(kubectl get pods -A | grep "datenlord-async" | awk 'NR==1{print $2}
 
 kubectl exec ${FIRST_NODE} -n ${NAMESPACE} -- apt-get update
 kubectl exec ${FIRST_NODE} -n ${NAMESPACE} -- apt-get install -y fio python3-pip
-kubectl exec ${FIRST_NODE} -n ${NAMESPACE} -- pip3 install matplotlib numpy fio-plot
+kubectl exec ${FIRST_NODE} -n ${NAMESPACE} -- pip3 install matplotlib numpy fio-plot==1.0.28
```
[the remark this answers is not in the document]
Comment (rogercloud) [reply scripts/datenlord_perf_test.sh:24]: It's a big change, let's take it into another PR. We can file an issue to track it.

File: src/async_fuse/fuse/file_system.rs [hunk 2]
```diff
@@ -283,4 +284,58 @@ pub trait FileSystem {
 
     /// Set fuse fd into `FileSystem`
     async fn set_fuse_fd(&self, fuse_fd: RawFd);
+
+    /// Stop all async tasks
+    fn stop_all_async_tasks(&self);
+}
+
+/// create channel of communication from async task to main loop
```
[the remark this answers is not in the document]
Comment (rogercloud) [reply src/async_fuse/fuse/file_system.rs:292]: What do you mean? Is the implementation of the `main loop` right?

File: src/async_fuse/fuse/session.rs [hunk 3]
```diff
@@ -105,9 +107,13 @@ impl Drop for FuseFd {
 impl Drop for Session {
     fn drop(&mut self) {
         futures::executor::block_on(async {
+            // join fuse request handling tasks.
             for join_handle in &self.tasks {
                 join_handle.abort();
             }
+            // stop and join async sub tasks
+            self.filesystem.stop_all_async_tasks();
+            self.fs_controller.join_all_async_tasks().await;
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/fuse/session.rs:116]: The `stop_all_async_tasks` changed the atomic flag in  filesystem, which is a shared data.
The fs_controller is owned by the session thread (main thread), `join_all_async_tasks` just consumes the joiners.
image

File: src/async_fuse/fuse/file_system.rs [hunk 4]
```diff
@@ -283,4 +284,58 @@ pub trait FileSystem {
...
+            task.await
+                .unwrap_or_else(|e| panic!("join async task error {e}"));
+        }
+    }
+    /// async read a result from async task
+    pub(crate) async fn recv_async_task_res(&mut self) -> anyhow::Result<()> {
+        if let Some(res) = self.async_res_receiver.recv().await {
+            res
+        } else {
+            // Only happens when channel sender destroyed,
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/fuse/file_system.rs:335]: Of course It dosn't panic logically, panic is just for debug some wrong logical change that when we didn't know the logic well. Just consider it as a unwrap_or_else("panic")

File: Cargo.toml [hunk 5]
```diff
@@ -63,6 +63,7 @@ protoc-grpcio = "3.0.0"
 
 [dev-dependencies]
 mock-etcd = { git = "https://github.com/datenlord/etcd-client", rev = "f697899" }
+mockall = "0.11.2"
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply Cargo.toml:66]: Ok.

File: src/async_fuse/fuse/file_system.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/fuse/file_system.rs:292]: Main loop refers to the session.

File: src/async_fuse/memfs/persist.rs [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+            }
+        }
+        left_more_dirty
+    }
+    /// Spawn the async persist task.
+    pub(crate) fn spawn(
+        s3_backend: Arc<S>,
+        fs_async_result_sender: FsAsyncResultSender,
+    ) -> (PersistHandle, JoinHandle<()>) {
+        let shared = Arc::new(PersistSharedState {
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/persist.rs:289]: Remain to be discussed.

File: src/async_fuse/memfs/persist.rs [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+    shared: Arc<PersistSharedState>,
+    /// S3 backend for persistence.
+    s3_backend: Arc<S>,
+}
+
+impl<S: S3BackEnd + Sync + Send + 'static> PersistTask<S> {
+    /// Called by persist task loop to take out one and persist to s3 each time.
+    #[inline]
+    async fn try_persist_one_dirty(&self, fs_async_result_sender: &FsAsyncResultSender) -> bool {
+        let mut to_persist = None;
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/persist.rs:232]: Remain to be discussed.

File: src/async_fuse/memfs/persist.rs [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+                    if i == PERSIST_RETRY_TIMES.overflow_sub(1) {
+                        // send err to main thread
+                        fs_async_result_sender
+                            .send(Err(anyhow::Error::from(e)))
+                            .await
+                            .unwrap_or_else(|e| {
+                                debug!("Failed to send error to main thread,{e}");
+                            });
+                    }
+                } else {
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/persist.rs:273]: Ok.

File: src/async_fuse/memfs/s3_node.rs [hunk 9]
```diff
@@ -264,34 +265,59 @@ impl<S: S3BackEnd + Send + Sync + 'static> S3Node<S> {
...
+                        name,
+                        "/".to_owned(),
+                        Arc::new(RwLock::new(attr)),
+                        data.new_s3_node_data_dir(),
+                        s3_backend,
+                        meta,
+                    );
 
-        Ok(root_node)
+                    Ok(root_node)
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/s3_node.rs:313]: Ok.

File: src/async_fuse/memfs/s3_metadata.rs [hunk 10]
```diff
@@ -492,14 +528,22 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
...
+        let (result, full_path, parent_ino) = {
             let mut cache = self.cache().write().await;
             let inode = cache.get_mut(&ino).unwrap_or_else(|| {
                 panic!(
                     "write() found fs is inconsistent, \
                      the inode ino={ino} is not in cache"
                 );
             });
+            let parent_ino = inode.get_parent_ino();
+            // let parent= cache.get_mut(&inode.get_parent_ino()).unwrap_or_else(|| {
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/s3_metadata.rs:540]: Ok.

File: src/async_fuse/fuse/file_system.rs [hunk 11]
```diff
@@ -283,4 +284,58 @@ pub trait FileSystem {
...
+    /// Stop all async tasks
+    fn stop_all_async_tasks(&self);
+}
+
+/// create channel of communication from async task to main loop
+pub(crate) fn new_fs_async_result_chan() -> (FsAsyncResultSender, FsAsyncResultReceiver) {
+    tokio::sync::mpsc::channel(10)
+}
+/// result of fs async result
+pub type FsAsyncResult = anyhow::Result<()>;
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/fuse/file_system.rs:297]: Ok.

File: src/async_fuse/memfs/persist.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/persist.rs:232]: TODO: change return to Result

File: src/async_fuse/memfs/persist.rs [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+                    DirEntry::new(
+                        name.clone(),
+                        Arc::new(RwLock::new(serial::serial_to_file_attr(fattr))),
+                    ),
+                )
+            })
+            .collect();
+        S3NodeData::Directory(dir_map)
+    }
+    // pub(crate) fn new() -> PersistDirContent {
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/persist.rs]: Ok.

File: paTools
Comment (xiaguan) [inline paTools:1]: This folder should not be submitted.

**Review by xiaguan — COMMENTED** [review]
Remove   `paTools`  .

File: .github/workflows/cron.yml [hunk 13]
```diff
@@ -170,11 +170,17 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Free Disk Space (Ubuntu)
-        uses: jlumbroso/free-disk-space@main
-        with:
-          # this might remove tools that are actually needed,
-          # when set to "true" but frees about 6 GB
-          tool-cache: true
+        run: |
```
Comment (rogercloud) [inline .github/workflows/cron.yml:173]: Why are we doing this? Do we have any disk limitation here?

File: src/async_fuse/fuse/file_system.rs [hunk 14]
```diff
@@ -283,4 +284,60 @@ pub trait FileSystem {
 
     /// Set fuse fd into `FileSystem`
     async fn set_fuse_fd(&self, fuse_fd: RawFd);
+
+    /// Stop all async tasks
+    fn stop_all_async_tasks(&self);
```
Comment (rogercloud) [inline src/async_fuse/fuse/file_system.rs:289]: Shall we move it into another trait? I don't think all the FileSystems have async tasks.

File: src/async_fuse/fuse/file_system.rs [hunk 15]
```diff
@@ -283,4 +284,60 @@ pub trait FileSystem {
...
+pub(crate) fn new_fs_async_result_chan() -> (FsAsyncResultSender, FsAsyncResultReceiver) {
+    tokio::sync::mpsc::channel(10)
+}
+/// result of fs async result
+pub type FsAsyncResult = DatenLordResult<()>;
+/// sender for async tasks to send msg(mainly refers to error) to session main loop
+pub type FsAsyncResultSender = tokio::sync::mpsc::Sender<FsAsyncResult>;
+/// receiver to receive msg from async tasks
+pub(crate) type FsAsyncResultReceiver = tokio::sync::mpsc::Receiver<FsAsyncResult>;
+/// Some state held by session to communicate with and control the fs.
```
Comment (rogercloud) [inline src/async_fuse/fuse/file_system.rs:302]: Please add an empty line here. It's a logical break for code reading.

File: src/async_fuse/fuse/session.rs [hunk 16]
```diff
@@ -150,77 +158,109 @@ impl Session {
             filesystem: Arc::new(fs),
             mount_path: mount_path.to_owned(),
             tasks: Vec::new(),
+            fs_controller,
         })
     }
 
     /// Run the FUSE session
-    #[allow(clippy::wildcard_enum_match_arm)] // nix::Errno is marked as non_exhaustive
+    #[allow(clippy::wildcard_enum_match_arm, clippy::restriction)] // nix::Errno is marked as non_exhaustive
```
Comment (rogercloud) [inline src/async_fuse/fuse/session.rs:166]: Why do we allow the "restriction" clippy here?

**Review by rogercloud — APPROVED** [review]

**Pull request merged**